core/src/main/java/org/elasticsearch/script/Script.java

```diff
@@ -19,272 +19,936 @@
+        }
+
+        @Override
+        public CompiledScript getCompiled(ScriptService service, ScriptContext context, ScriptBinding binding) {
+            return service.getInlineScript(context, binding, this);
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeString(lang);
+            out.writeString(code);
+            out.writeInt(options.size());
+
+            for (Map.Entry<String, String> option : options.entrySet()) {
+                out.writeString(option.getKey());
+                out.writeString(option.getValue());
+            }
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.field(ScriptField.LANG.getPreferredName(), lang);
+
+            String content = options.get(CONTENT_TYPE_OPTION);
+
+            if (content != null && content.equals(builder.contentType().mediaType())) {
+                builder.rawField(INLINE.parse.getPreferredName(), new BytesArray(code));
             } else {
-                throw new ElasticsearchParseException("unexpected field [{}]", cfn);
+                builder.field(INLINE.parse.getPreferredName(), code);
             }
+
+            return builder;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+
+            InlineScriptLookup that = (InlineScriptLookup)o;
+
+            if (!lang.equals(that.lang)) return false;
+            if (!code.equals(that.code)) return false;
+            return options.equals(that.options);
+
         }
-        if (script == null) {
-            throw new ElasticsearchParseException("expected one of [{}], [{}] or [{}] fields, but found none",
-                    ScriptType.INLINE.getParseField() .getPreferredName(), ScriptType.FILE.getParseField().getPreferredName(),
-                    ScriptType.STORED.getParseField() .getPreferredName());
+
+        @Override
+        public int hashCode() {
+            int result = lang.hashCode();
+            result = 31 * result + code.hashCode();
+            result = 31 * result + options.hashCode();
+            return result;
+        }
+
+        @Override
+        public String toString() {
+            return "InlineScriptLookup{" +
+                "lang='" + lang + '\'' +
+                ", code='" + code + '\'' +
+                ", options=" + options +
+                '}';
         }
-        return new Script(script, type, lang, params, contentType);
     }
 
-    @Override
-    public int hashCode() {
-        return Objects.hash(lang, params, script, type, contentType);
+    public interface ScriptBinding {
+        Object compile(ScriptEngineService engine, String id, String code, Map<String, String> options);
```

Comment:
Should this be `ScriptBinding<T>` where `T` is the interface that we expect the script to implement? Then `compile` can return `T`?